File: practicas/practicas-cpp/CajeroBD/src/DatabaseManager/DatabaseManager.cpp
//DatabaseManager.cpp
#include "DatabaseManager.h"
#include <iostream>
#include <string>
#include <fstream>
#include <unistd.h>

// Constructor: Inicializa la conexión MySQL
DatabaseManager::DatabaseManager() : conn(nullptr) {
    loadProperties();
    host = this->properties["db.host"];
    user = this->properties["db.user"];
    pass = this->properties["db.pass"];
    db = this->properties["db.name"];
    std::string portStr = properties["db.port"];
    size_t idx;
    try {
        port = std::stoi(portStr, &idx);
        if (idx != portStr.size()) {
            throw std::invalid_argument("Invalid port number");
        }
    } catch (const std::invalid_argument& e) {
        std::cerr << "Error: invalid port number in database.properties" << std::endl;
        // Maneja el error aquí, por ejemplo, estableciendo un valor por defecto para el puerto
        port = 3306; // valor por defecto
    }

    conn = mysql_init(nullptr);
    if (!conn) {
        std::cerr << "Error: mysql_init falló en DatabaseManager.\n";
    }
}

// Destructor: Cierra la conexión MySQL
DatabaseManager::~DatabaseManager() {
    if (conn) {
        mysql_close(conn);
        std::cout << "\nConexión a la base de datos cerrada por DatabaseManager.\n";
    }
}

// void DatabaseManager::loadProperties() {
//     // Comprobación de la ruta del archivo
//     if (!std::ifstream("../../database.properties").good()) {
//         std::cerr << "Error: no se pudo encontrar el archivo database.properties" << std::endl;
//         return;
//     }

//     // Comprobación de permisos de acceso
//     if (access("../../database.properties", R_OK) != 0) {
//         std::cerr << "Error: no se tienen permisos de acceso para leer el archivo database.properties" << std::endl;
//         return;
//     }

//     // Comprobación del formato del archivo
//     std::ifstream propertiesFile("../../database.properties");
//     if (!propertiesFile.is_open() || !propertiesFile.good()) {
//         std::cerr << "Error: no se pudo abrir o leer el archivo database.properties" << std::endl;
//         return;
//     }

//     // Comprobación de la estructura del archivo
//     std::string line;
//     while (std::getline(propertiesFile, line)) {
//         if (line.empty() || line.find('=') == std::string::npos) {
//             std::cerr << "Error: el archivo database.properties tiene un formato incorrecto" << std::endl;
//             return;
//         }
//         // ...
//     }

//     // Comprobación de la existencia de propiedades
//     if (properties.empty()) {
//         std::cerr << "Error: no se encontraron propiedades en el archivo database.properties" << std::endl;
//         return;
//     }
// }

void DatabaseManager::loadProperties() {
    std::ifstream propertiesFile("../../database.properties");
    if (propertiesFile.is_open()) {
        std::string line;
        while (std::getline(propertiesFile, line)) {
            size_t igual = line.find('=');
            if (igual != std::string::npos) {
                std::string clave = line.substr(0, igual);
                std::string valor = line.substr(igual + 1);
                properties[clave] = valor;
            }
        }
    } else {
        std::cerr << "Error al abrir el archivo de propiedades." << std::endl;
    }
}

// Conecta a la base de datos con los parámetros definidos en el header
bool DatabaseManager::connect() {
    if (!mysql_real_connect(conn, host.c_str(), user.c_str(), pass.c_str(), db.c_str(), port, NULL, 0)) {
        std::cerr << "Error al conectar a la base de datos: " << mysql_error(conn) << std::endl;
        return false;
    }
    std::cout << "Conexión exitosa a la base de datos '" << db << "'\n";
    return true;
}

// Ejecuta una consulta SQL y retorna true si no hay error
bool DatabaseManager::validateteQuery(const std::string& query) {
    if (mysql_query(conn, query.c_str())) {
        std::cerr << "Error al ejecutar la consulta: " << mysql_error(conn) << std::endl;
        return false;
    }
    return true;
}

// Inserta datos en la tabla indicada, usando columnas y valores dinámicos
// Uso: insertData("usuario", {"ID", "Nombre", "Apellido"}, {"123", "Juan", "Pérez"});
bool DatabaseManager::insertData(const std::string& tableName, 
                                const std::vector<std::string>& columns, 
                                const std::vector<std::string>& values) {
    if (columns.size() != values.size() || columns.empty()) {
        std::cerr << "Error: columnas y valores no coinciden o están vacíos.\n";
        return false;
    }
    std::string query = "INSERT INTO " + tableName + " (";
    for (size_t i = 0; i < columns.size(); ++i) {
        query += columns[i];
        if (i < columns.size() - 1) query += ", ";
    }
    query += ") VALUES (";
    for (size_t i = 0; i < values.size(); ++i) {
        query += "'" + values[i] + "'";
        if (i < values.size() - 1) query += ", ";
    }
    query += ")";
    std::cout << "Ejecutando: " << query << std::endl;
    if (validateteQuery(query)) {

        std::cout << "Datos insertados correctamente en " + tableName + ".\n";
        return true;
    } else {
        std::cerr << "Fallo al insertar datos.\n";
        return false;
    }
}

// Lee y muestra todos los datos de la tabla indicada
// Uso: readData("usuario");
void DatabaseManager::readData(const std::string& tableName) {
    std::string query = "SELECT * FROM " + tableName;
    if (validateteQuery(query)) {
        MYSQL_RES* res = mysql_store_result(conn);
        if (res) {
            MYSQL_ROW row;
            int num_fields = mysql_num_fields(res);
            MYSQL_FIELD *fields;
            fields = mysql_fetch_fields(res);

            std::cout << "\n--- Datos de la tabla '" << tableName << "' ---\n";
            
            while ((row = mysql_fetch_row(res))) {
                std::cout << "-------------------------------------\n";
                for (int i = 0; i < num_fields; i++) {                    
                    std::cout << fields[i].name << ": " << (row[i] ? row[i] : "NULL") << "\n";
                }
                std::cout << "\n";
            }
            std::cout << "--- Fin de datos de la tabla '" << tableName << "' ---\n";
            mysql_free_result(res);
        } else {
            std::cerr << "No se pudo obtener el resultado de la consulta: " << mysql_error(conn) << std::endl;
        }
    } else {
        std::cerr << "Fallo al leer datos.\n";
    }
}

// Ejecuta cualquier consulta SQL (SELECT, INSERT, UPDATE, etc.) y muestra resultados si los hay
// Uso: executeQuery("SELECT * FROM usuario WHERE ID='123'");
void DatabaseManager::executeQuery(const std::string& query) {
    if (mysql_query(conn, query.c_str())) {
        std::cerr << "Error al ejecutar la consulta SELECT: " << mysql_error(conn) << std::endl;
        return;
    }

    MYSQL_RES* res = mysql_store_result(conn);
    if (res) {
        MYSQL_ROW row;
        int num_fields = mysql_num_fields(res);
        MYSQL_FIELD *fields = mysql_fetch_fields(res);

        std::cout << "\n--- Resultados de la Consulta Personalizada ---\n";
        if (mysql_num_rows(res) == 0) {
            std::cout << "No se encontraron resultados para la consulta.\n";
        } else {
            while ((row = mysql_fetch_row(res))) {
                std::cout << "-------------------------------------\n";
                for (int i = 0; i < num_fields; i++) {                    
                    std::cout << fields[i].name << ": " << (row[i] ? row[i] : "NULL") << "\n";
                }
                std::cout << "\n";
            }
        }
        std::cout << "--- Fin de resultados personalizados ---\n";
        mysql_free_result(res);
    } else if (mysql_field_count(conn) == 0) {                
        std::cout << "Consulta ejecutada con éxito (no es una consulta SELECT que devuelva resultados).\n";
    } else {
        std::cerr << "No se pudo obtener el resultado de la consulta: " << mysql_error(conn) << std::endl;
    }
}

std::string DatabaseManager::executeScalar(const std::string& query) {
    if (mysql_query(conn, query.c_str())) {
        std::cerr << "Error al ejecutar la consulta: " << mysql_error(conn) << std::endl;
        return "";
    }
    MYSQL_RES* res = mysql_store_result(conn);
    if (res) {
        MYSQL_ROW row = mysql_fetch_row(res);
        std::string result = (row && row[0]) ? row[0] : "";
        mysql_free_result(res);
        return result;
    }
    return "";
}


// Actualiza datos en la tabla indicada, usando columnas a modificar y condiciones dinámicas
// Uso: updateData("usuario", {{"Nombre", "Pedro"}}, {{"ID", "123"}});
bool DatabaseManager::updateData(const std::string& tableName, 
                                const std::map<std::string, std::string>& setColumns, 
                                const std::map<std::string, std::string>& whereConditions) {
    if (setColumns.empty() || whereConditions.empty()) {
        std::cerr << "Error: columnas a actualizar o condiciones vacías.\n";
        return false;
    }
    std::string query = "UPDATE " + tableName + " SET ";
    size_t count = 0;
    for (const auto& pair : setColumns) {
        query += pair.first + " = '" + pair.second + "'";
        if (++count < setColumns.size()) query += ", ";
    }
    query += " WHERE ";
    count = 0;
    for (const auto& pair : whereConditions) {
        query += pair.first + " = '" + pair.second + "'";
        if (++count < whereConditions.size()) query += " AND ";
    }
    std::cout << "Ejecutando: " << query << std::endl;
    if (validateteQuery(query)) {
        std::cout << "Datos actualizados correctamente en " + tableName + ".\n";
        return true;
    } else {
        std::cerr << "Fallo al actualizar datos.\n";
        return false;
    }
}

// Elimina datos de la tabla indicada según condiciones dinámicas
// Uso: deleteData("usuario", {{"ID", "123"}});
bool DatabaseManager::deleteData(const std::string& tableName, 
                                const std::map<std::string, std::string>& whereConditions) {
    if (whereConditions.empty()) {
        std::cerr << "Error: condiciones vacías para eliminar.\n";
        return false;
    }
    std::string query = "DELETE FROM " + tableName + " WHERE ";
    size_t count = 0;
    for (const auto& pair : whereConditions) {
        query += pair.first + " = '" + pair.second + "'";
        if (++count < whereConditions.size()) query += " AND ";
    }
    std::cout << "Ejecutando: " << query << std::endl;
    if (validateteQuery(query)) {
        std::cout << "Datos eliminados correctamente de " + tableName + ".\n";
        return true;
    } else {
        std::cerr << "Fallo al eliminar datos.\n";
        return false;
    }
}

// Verifica si un registro existe según una consulta dada
// Uso: existeRegistro("SELECT * FROM usuario WHERE ID='123'");
bool DatabaseManager::existsData(const std::string& query) {
    if (mysql_query(conn, query.c_str())) return false;
    MYSQL_RES* res = mysql_store_result(conn);
    bool existe = (res && mysql_num_rows(res) > 0);
    if (res) mysql_free_result(res);
    return existe;
}

std::vector<std::map<std::string, std::string>> DatabaseManager::fetchQuery(const std::string& query) {
    std::vector<std::map<std::string, std::string>> resultados;
    if (mysql_query(conn, query.c_str()) == 0) {
        MYSQL_RES* res = mysql_store_result(conn);
        if (res) {
            int num_fields = mysql_num_fields(res);
            MYSQL_FIELD* fields = mysql_fetch_fields(res);
            MYSQL_ROW row;
            while ((row = mysql_fetch_row(res))) {
                std::map<std::string, std::string> registro;
                for (int i = 0; i < num_fields; ++i) {
                    registro[fields[i].name] = row[i] ? row[i] : "";
                }
                resultados.push_back(registro);
            }
            mysql_free_result(res);
        }
    }
    return resultados;
}

File: practicas/practicas-java/Cajero/src/com/managment/database/DatabaseManager.java
package com.managment.database;

import java.io.InputStream;
import java.sql.*;
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * DatabaseManager.java
 * Clase para gestionar la conexión a la base de datos y realizar operaciones CRUD.
 * Adaptada para entorno GUI (sin impresión directa, devuelve datos como String o listas)
 * @autor Ektor Ormaetxea V
 */
public class DatabaseManager implements AutoCloseable {
    private static final Logger LOGGER = Logger.getLogger(DatabaseManager.class.getName());

    private Connection connection;

    private String DB_HOST;
    private String DB_USER;
    private String DB_PASS;
    private String DB_NAME;
    private int DB_PORT;
    private final String DB_URL;

    /**
     * Constructor que carga las propiedades de configuración y prepara la URL de conexión.
     */
    public DatabaseManager() {
        loadProperties();
        this.DB_URL = "jdbc:mysql://" + DB_HOST + ":" + DB_PORT + "/" + DB_NAME + "?useSSL=false&serverTimezone=UTC";
    }

    /**
     * Carga las propiedades desde config.properties.
     */
    private void loadProperties() {
        Properties prop = new Properties();
        try (InputStream input = getClass().getClassLoader().getResourceAsStream("config.properties")) {
            if (input == null) throw new RuntimeException("No se encontró el archivo de configuración.");
            prop.load(input);
            DB_HOST = prop.getProperty("db.host");
            DB_USER = prop.getProperty("db.user");
            DB_PASS = prop.getProperty("db.pass");
            DB_NAME = prop.getProperty("db.name");
            DB_PORT = Integer.parseInt(prop.getProperty("db.port"));
        } catch (Exception ex) {
            LOGGER.log(Level.SEVERE, "Error al cargar las propiedades: " + ex.getMessage(), ex);
            throw new RuntimeException("Error al cargar configuración", ex);
        }
    }

    /**
     * Establece conexión con la base de datos.
     * @return true si la conexión fue exitosa.
     */
    public boolean connect() {
        try {
            connection = DriverManager.getConnection(DB_URL, DB_USER, DB_PASS);
            LOGGER.info("Conexión exitosa a la base de datos");
            return true;
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Error al conectar: " + e.getMessage(), e);
            return false;
        }
    }

    /**
     * Lee todos los datos de una tabla.
     * @param tableName Nombre de la tabla.
     * @return Lista de filas (mapas columna-valor).
     */
    public List<Map<String, String>> readData(String tableName) {
        return fetchQuery("SELECT * FROM " + tableName);
    }

    /**
     * Lee y devuelve los datos como texto.
     * @param tableName Nombre de la tabla.
     * @return Resultados como texto.
     */
    public String readDataAsString(String tableName) {
        return printQuery(readData(tableName));
    }

    /**
     * Convierte una lista de mapas en texto legible.
     * @param resultados Lista de resultados.
     * @return Cadena formateada.
     */
    public String printQuery(List<Map<String, String>> resultados) {
        if (resultados.isEmpty()) return "No hay resultados";

        StringBuilder result = new StringBuilder();
        for (Map<String, String> registro : resultados) {
            for (Map.Entry<String, String> par : registro.entrySet()) {
                result.append(par.getKey()).append(": ").append(par.getValue()).append("\n");
            }
            result.append("-------------------------\n");
        }
        return result.toString();
    }

    /**
     * Ejecuta una consulta SELECT y devuelve los resultados.
     * @param query Consulta SQL.
     * @return Lista de resultados.
     */
    public List<Map<String, String>> fetchQuery(String query) {
        List<Map<String, String>> resultados = new ArrayList<>();

        try (Statement stmt = connection.createStatement();
             ResultSet rs = stmt.executeQuery(query)) {

            ResultSetMetaData rsmd = rs.getMetaData();
            int numColumns = rsmd.getColumnCount();

            while (rs.next()) {
                Map<String, String> fila = new LinkedHashMap<>();
                for (int i = 1; i <= numColumns; i++) {
                    String nombreColumna = rsmd.getColumnLabel(i);
                    fila.put(nombreColumna, rs.getString(i));
                }
                resultados.add(fila);
            }
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Error en fetchQuery: " + e.getMessage(), e);
        }

        return resultados;
    }

    /**
     * Inserta datos usando Map columna-valor.
     * @param tableName Tabla de destino.
     * @param columnValues Mapa columna-valor.
     * @return true si se insertó correctamente.
     */
    public boolean insertData(String tableName, Map<String, String> columnValues) {
        if (columnValues == null || columnValues.isEmpty()) return false;

        StringBuilder query = new StringBuilder("INSERT INTO ").append(tableName).append(" (");
        StringBuilder placeholders = new StringBuilder();

        List<String> keys = new ArrayList<>(columnValues.keySet());
        for (int i = 0; i < keys.size(); i++) {
            query.append(keys.get(i));
            placeholders.append("?");
            if (i < keys.size() - 1) {
                query.append(", ");
                placeholders.append(", ");
            }
        }
        query.append(") VALUES (").append(placeholders).append(")");

        LOGGER.info("Preparando: " + query.toString());

        try (PreparedStatement pstmt = connection.prepareStatement(query.toString())) {
            int index = 1;
            for (String key : keys) {
                pstmt.setString(index++, columnValues.get(key));
            }
            int rowsAffected = pstmt.executeUpdate();
            LOGGER.info("Datos insertados correctamente en " + tableName + ". Filas afectadas: " + rowsAffected);
            return rowsAffected > 0;
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Fallo al insertar datos: " + e.getMessage(), e);
            return false;
        }
    }

    /**
     * Inserta datos usando listas columna-valor.
     * @param tableName Tabla destino.
     * @param columns Lista de columnas.
     * @param values Lista de valores.
     * @return true si se insertó correctamente.
     */
    public boolean insertDataByList(String tableName, List<String> columns, List<String> values) {
        if (columns.size() != values.size() || columns.isEmpty()) {
            LOGGER.severe("Error: columnas y valores no coinciden o están vacíos.");
            return false;
        }

        StringBuilder query = new StringBuilder("INSERT INTO ").append(tableName).append(" (");
        for (int i = 0; i < columns.size(); i++) {
            query.append(columns.get(i));
            if (i < columns.size() - 1) query.append(", ");
        }
        query.append(") VALUES (");
        for (int i = 0; i < values.size(); i++) {
            query.append("?");
            if (i < values.size() - 1) query.append(", ");
        }
        query.append(")");

        LOGGER.info("Preparando: " + query.toString());

        try (PreparedStatement pstmt = connection.prepareStatement(query.toString())) {
            for (int i = 0; i < values.size(); i++) {
                pstmt.setString(i + 1, values.get(i));
            }
            int rowsAffected = pstmt.executeUpdate();
            LOGGER.info("Datos insertados correctamente en " + tableName + ". Filas afectadas: " + rowsAffected);
            return rowsAffected > 0;
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Fallo al insertar datos: " + e.getMessage(), e);
            return false;
        }
    }

    /**
     * Elimina datos según condiciones.
     * @param tableName Tabla objetivo.
     * @param whereConditions Condiciones WHERE.
     * @return true si se eliminó al menos una fila.
     */
    public boolean deleteData(String tableName, Map<String, String> whereConditions) {
        if (whereConditions == null || whereConditions.isEmpty()) return false;

        StringBuilder query = new StringBuilder("DELETE FROM ").append(tableName).append(" WHERE ");
        List<String> keys = new ArrayList<>(whereConditions.keySet());

        for (int i = 0; i < keys.size(); i++) {
            query.append(keys.get(i)).append(" = ?");
            if (i < keys.size() - 1) query.append(" AND ");
        }

        try (PreparedStatement pstmt = connection.prepareStatement(query.toString())) {
            int index = 1;
            for (String key : keys) {
                pstmt.setString(index++, whereConditions.get(key));
            }
            return pstmt.executeUpdate() > 0;
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Error en deleteData: " + e.getMessage(), e);
            return false;
        }
    }

    /**
     * Actualiza datos con condiciones.
     * @param tableName Nombre de la tabla.
     * @param setColumns Nuevos valores.
     * @param whereConditions Condiciones WHERE.
     * @return true si se actualizó al menos una fila.
     */
    public boolean updateData(String tableName, Map<String, String> setColumns, Map<String, String> whereConditions) {
        if (setColumns == null || setColumns.isEmpty() || whereConditions == null || whereConditions.isEmpty()) {
            return false;
        }

        StringBuilder query = new StringBuilder("UPDATE ").append(tableName).append(" SET ");
        List<String> setKeys = new ArrayList<>(setColumns.keySet());
        for (int i = 0; i < setKeys.size(); i++) {
            query.append(setKeys.get(i)).append(" = ?");
            if (i < setKeys.size() - 1) query.append(", ");
        }

        query.append(" WHERE ");
        List<String> whereKeys = new ArrayList<>(whereConditions.keySet());
        for (int i = 0; i < whereKeys.size(); i++) {
            query.append(whereKeys.get(i)).append(" = ?");
            if (i < whereKeys.size() - 1) query.append(" AND ");
        }

        LOGGER.info("Ejecutando UPDATE: " + query.toString());

        try (PreparedStatement pstmt = connection.prepareStatement(query.toString())) {
            int index = 1;
            for (String key : setKeys) {
                pstmt.setString(index++, setColumns.get(key));
            }
            for (String key : whereKeys) {
                pstmt.setString(index++, whereConditions.get(key));
            }
            int rowsAffected = pstmt.executeUpdate();
            LOGGER.info("Filas actualizadas: " + rowsAffected);
            return rowsAffected > 0;
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Error en updateData: " + e.getMessage(), e);
            return false;
        }
    }

    /**
     * Ejecuta cualquier consulta SQL y devuelve un resultado como String.
     * @param query Consulta a ejecutar.
     * @return Resultado o mensaje.
     */
    public String executeQuery(String query) {
        String trimmedQuery = query.trim().toUpperCase();
        boolean isSelect = trimmedQuery.startsWith("SELECT");

        if (isSelect) {
            List<Map<String, String>> resultados = fetchQuery(query);
            return printQuery(resultados);
        } else {
            try (Statement stmt = connection.createStatement()) {
                int rowsAffected = stmt.executeUpdate(query);
                return "Consulta ejecutada correctamente. Filas afectadas: " + rowsAffected;
            } catch (SQLException e) {
                LOGGER.log(Level.SEVERE, "Error al ejecutar la consulta: " + e.getMessage(), e);
                return "Error al ejecutar la consulta: " + e.getMessage();
            }
        }
    }

    /**
     * Ejecuta una consulta que devuelve solo un valor.
     * @param query Consulta tipo escalar.
     * @return Primer valor encontrado.
     */
    public String executeScalar(String query) {
        try (Statement stmt = connection.createStatement();
             ResultSet rs = stmt.executeQuery(query)) {
            if (rs.next()) {
                return rs.getString(1);
            }
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Error al ejecutar la consulta escalar: " + e.getMessage(), e);
        }
        return "";
    }

    /**
     * Cierra la conexión con la base de datos.
     */
    @Override
    public void close() throws SQLException {
        if (connection != null && !connection.isClosed()) {
            connection.close();
            LOGGER.info("Conexión cerrada.");
        }
    }
}
